blocks/certifier.go

```diff
@@ -0,0 +1,394 @@
+package blocks
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/spacemeshos/ed25519"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/hare"
+	"github.com/spacemeshos/go-spacemesh/hare/eligibility"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/p2p/pubsub"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/layers"
+)
+
+var (
+	errKnownCertMsg   = errors.New("known certify msg")
+	errMsgTooOld      = errors.New("cert msg too old")
+	errInvalidCertMsg = errors.New("invalid cert msg")
+	errUnexpectedMsg  = errors.New("unexpected lid/bid")
+	errInternal       = errors.New("internal")
+)
+
+// CertConfig is the config for Certifier.
+type CertConfig struct {
+	CommitteeSize         int
+	CertifyThreshold      int
+	SignatureWaitDuration time.Duration
+	NumLayersToKeep       uint32
+}
+
+func defaultCertConfig() CertConfig {
+	return CertConfig{
+		CommitteeSize:         10,
+		CertifyThreshold:      6,
+		SignatureWaitDuration: time.Second * 10,
+		NumLayersToKeep:       5,
+	}
+}
+
+// CertifierOpt for configuring Certifier.
+type CertifierOpt func(*Certifier)
+
+// WithCertContext modifies parent context for Certifier.
+func WithCertContext(ctx context.Context) CertifierOpt {
+	return func(c *Certifier) {
+		c.ctx = ctx
+	}
+}
+
+// WithCertConfig defines cfg for Certifier.
+func WithCertConfig(cfg CertConfig) CertifierOpt {
+	return func(c *Certifier) {
+		c.cfg = cfg
+	}
+}
+
+// WithCertifierLogger defines logger for Certifier.
+func WithCertifierLogger(logger log.Log) CertifierOpt {
+	return func(c *Certifier) {
+		c.logger = logger
+	}
+}
+
+type certInfo struct {
+	bid              types.BlockID
+	deadline         time.Time
+	done             bool
+	totalEligibility uint16
+	signatures       []types.CertifyMessage
+}
+
+type certState int
+
+const (
+	unknown certState = iota
+	pending
+	certified
+	expired
+)
+
+// Certifier collects enough CertifyMessage for a given hare output and generate certificate.
+type Certifier struct {
+	logger log.Log
+	cfg    CertConfig
+	once   sync.Once
+	eg     errgroup.Group
+	ctx    context.Context
+	cancel func()
+
+	db         *sql.Database
+	oracle     hare.Rolacle
+	nodeID     types.NodeID
+	signer     *signing.EdSigner
+	publisher  pubsub.Publisher
+	layerClock layerClock
+
+	mu       sync.Mutex
+	certMsgs map[types.LayerID]*certInfo
```

Comment:
rename to `certs`

Reply:
but they are not certs. i used the name `cert` for the actual certificate. i don't want my readers to confuse the two (CertifyMessage and Certificate)

Reply:
renamed to certifyMsgs to be clear

Reply:
fair point, that works